File: scripts/lib/output-sql.ts
import type { ExtractedDevelopment } from './types.js';
import { generateSlug, sqlString, sqlBool, sqlInt, sqlDate } from './utils.js';

export function generateSQL(
  developments: ExtractedDevelopment[],
  operatorName: string,
  operatorWebsite?: string
): string {
  const operatorSlug = generateSlug(operatorName);
  const now = new Date().toISOString().split('T')[0];

  // Only include MEDIUM+ confidence
  const included = developments.filter(d => d.confidence.level !== 'LOW');
  const excluded = developments.filter(d => d.confidence.level === 'LOW');

  // Collect unique asset owners (that differ from operator)
  const assetOwners = new Set<string>();
  for (const dev of included) {
    if (dev.assetOwner && dev.assetOwner.toLowerCase() !== operatorName.toLowerCase()) {
      assetOwners.add(dev.assetOwner);
    }
  }

  const lines: string[] = [];

  // Header
  lines.push(`-- ============================================================================`);
  lines.push(`-- BTR Directory Data Upload: ${operatorName}`);
  lines.push(`-- Generated: ${now}`);
  lines.push(`-- REVIEW STATUS: PENDING MANUAL CHECK`);
  lines.push(`-- Total developments found: ${developments.length}`);
  lines.push(`-- Included in SQL (MEDIUM+ confidence): ${included.length}`);
  lines.push(`-- Excluded (LOW confidence, review CSV): ${excluded.length}`);
  lines.push(`-- ============================================================================`);
  lines.push('');

  // Step 1: Insert operator
  lines.push(`-- Step 1: Insert operator (if not already in database)`);
  lines.push(`INSERT INTO operators (name, slug, website, description)`);
  lines.push(`SELECT ${sqlString(operatorName)}, ${sqlString(operatorSlug)}, ${sqlString(operatorWebsite)}, ${sqlString(`BTR operator`)}`);
  lines.push(`WHERE NOT EXISTS (SELECT 1 FROM operators WHERE slug = ${sqlString(operatorSlug)});`);
  lines.push('');

  // Step 2: Insert asset owners
  if (assetOwners.size > 0) {
    lines.push(`-- Step 2: Insert asset owners (if not already in database)`);
    for (const aoName of assetOwners) {
      const aoSlug = generateSlug(aoName);
      lines.push(`INSERT INTO asset_owners (name, slug)`);
      lines.push(`SELECT ${sqlString(aoName)}, ${sqlString(aoSlug)}`);
      lines.push(`WHERE NOT EXISTS (SELECT 1 FROM asset_owners WHERE slug = ${sqlString(aoSlug)});`);
      lines.push('');
    }
  }

  // Step 3: Insert developments using DO block
  lines.push(`-- Step 3: Insert developments`);
  lines.push(`DO $$`);
  lines.push(`DECLARE`);
  lines.push(`  op_id UUID;`);
  if (assetOwners.size > 0) {
    lines.push(`  ao_id UUID;`);
  }
  lines.push(`BEGIN`);
  lines.push(`  SELECT id INTO op_id FROM operators WHERE slug = ${sqlString(operatorSlug)};`);
  lines.push('');

  for (const dev of included) {
    const sameOwner = !dev.assetOwner || dev.assetOwner.toLowerCase() === operatorName.toLowerCase();
    const aoSlug = sameOwner ? null : generateSlug(dev.assetOwner);

    lines.push(`  -- DEVELOPMENT: ${dev.name}`);
    lines.push(`  -- Confidence: ${dev.confidence.level} (${dev.confidence.overall})`);
    lines.push(`  -- Sources: ${dev.sourceUrls.join(', ')}`);
    if (dev.extractionNotes.length > 0) {
      lines.push(`  -- Notes: ${dev.extractionNotes.join('; ')}`);
    }

    if (aoSlug) {
      lines.push(`  SELECT id INTO ao_id FROM asset_owners WHERE slug = ${sqlString(aoSlug)};`);
    }

    lines.push(`  INSERT INTO developments (`);
    lines.push(`    name, slug, development_type, operator_id, asset_owner_id,`);
    lines.push(`    area, region, postcode,`);
    lines.push(`    number_of_units, status, completion_date,`);
    lines.push(`    description, website_url,`);
    lines.push(`    amenity_gym, amenity_pool, amenity_coworking, amenity_concierge,`);
    lines.push(`    amenity_cinema, amenity_roof_terrace, amenity_bike_storage,`);
    lines.push(`    amenity_pet_facilities, amenity_ev_charging, amenity_parcel_room,`);
    lines.push(`    amenity_guest_suites, amenity_playground,`);
    lines.push(`    pets_allowed, is_published, flagged_for_review`);
    lines.push(`  )`);
    lines.push(`  SELECT`);
    lines.push(`    ${sqlString(dev.name)}, ${sqlString(dev.slug)}, ${sqlString(dev.developmentType)}, op_id, ${sameOwner ? 'op_id' : 'ao_id'},`);
    lines.push(`    ${sqlString(dev.area)}, ${sqlString(dev.region)}, ${sqlString(dev.postcode)},`);
    lines.push(`    ${sqlInt(dev.numberOfUnits)}, ${sqlString(dev.status)}, ${sqlDate(dev.completionDate)},`);
    lines.push(`    ${sqlString(dev.description.substring(0, 500))}, ${sqlString(dev.websiteUrl)},`);
    lines.push(`    ${sqlBool(dev.amenities.amenity_gym)}, ${sqlBool(dev.amenities.amenity_pool)}, ${sqlBool(dev.amenities.amenity_coworking)}, ${sqlBool(dev.amenities.amenity_concierge)},`);
    lines.push(`    ${sqlBool(dev.amenities.amenity_cinema)}, ${sqlBool(dev.amenities.amenity_roof_terrace)}, ${sqlBool(dev.amenities.amenity_bike_storage)},`);
    lines.push(`    ${sqlBool(dev.amenities.amenity_pet_facilities)}, ${sqlBool(dev.amenities.amenity_ev_charging)}, ${sqlBool(dev.amenities.amenity_parcel_room)},`);
    lines.push(`    ${sqlBool(dev.amenities.amenity_guest_suites)}, ${sqlBool(dev.amenities.amenity_playground)},`);
    lines.push(`    ${sqlBool(dev.petsAllowed)}, true, ${sqlBool(dev.confidence.level === 'MEDIUM')}`);
    lines.push(`  WHERE NOT EXISTS (`);
    lines.push(`    SELECT 1 FROM developments WHERE slug = ${sqlString(dev.slug)}`);
    lines.push(`  );`);
    lines.push('');
  }

  lines.push(`END $$;`);
  lines.push('');

  // Summary comment
  lines.push(`-- ============================================================================`);
  lines.push(`-- Summary`);
  lines.push(`-- ============================================================================`);
  for (const dev of included) {
    const flag = dev.confidence.level === 'MEDIUM' ? ' [FLAGGED FOR REVIEW]' : '';
    lines.push(`-- ${dev.confidence.level}: ${dev.name} (${dev.area || 'unknown area'})${flag}`);
  }
  if (excluded.length > 0) {
    lines.push(`--`);
    lines.push(`-- LOW confidence (not included - review CSV):`);
    for (const dev of excluded) {
      lines.push(`--   ${dev.name} (${dev.extractionNotes.join('; ')})`);
    }
  }

  return lines.join('\n');
}

File: scripts/discover/output_sql.py
import re
from pathlib import Path

from models import Confidence, DiscoveredDevelopment
from deduplicator import generate_slug


def generate_sql_inserts(
    developments: list[DiscoveredDevelopment],
    date_str: str,
    output_dir: Path,
) -> Path:
    """
    Generate SQL INSERT statements for new developments.
    Only includes MEDIUM+ confidence NEW developments.
    Uses WHERE NOT EXISTS for dedup safety.
    """
    filename = f"discovery_upload_{date_str}.sql"
    filepath = output_dir / filename

    # Filter: only NEW + MEDIUM or HIGH confidence
    eligible = [
        d for d in developments
        if d.is_new and d.confidence in (Confidence.HIGH, Confidence.MEDIUM)
    ]

    lines = []
    lines.append("-- BTR Discovery Upload SQL")
    lines.append(f"-- Generated: {date_str}")
    lines.append(f"-- Developments: {len(eligible)} (MEDIUM+ confidence, NEW only)")
    lines.append("-- REVIEW CAREFULLY BEFORE EXECUTING")
    lines.append("--")
    lines.append("-- This file uses WHERE NOT EXISTS to prevent duplicate inserts.")
    lines.append("-- It is safe to run multiple times.")
    lines.append("")

    if not eligible:
        lines.append("-- No eligible developments found (need MEDIUM+ confidence + NEW status)")
        with open(filepath, "w", encoding="utf-8") as f:
            f.write("\n".join(lines))
        return filepath

    # Step 1: Collect unique operators and asset owners
    operators = set()
    asset_owners = set()
    for dev in eligible:
        if dev.operator_name:
            operators.add(dev.operator_name)
        if dev.asset_owner_name:
            asset_owners.add(dev.asset_owner_name)

    # Step 2: Insert operators
    if operators:
        lines.append("-- ============================================================")
        lines.append("-- Step 1: Insert operators (if they don't already exist)")
        lines.append("-- ============================================================")
        lines.append("")
        for op_name in sorted(operators):
            op_slug = generate_slug(op_name)
            lines.append(f"INSERT INTO operators (name, slug)")
            lines.append(f"SELECT {_sql_str(op_name)}, {_sql_str(op_slug)}")
            lines.append(f"WHERE NOT EXISTS (SELECT 1 FROM operators WHERE slug = {_sql_str(op_slug)});")
            lines.append("")

    # Step 3: Insert asset owners
    if asset_owners:
        lines.append("-- ============================================================")
        lines.append("-- Step 2: Insert asset owners (if they don't already exist)")
        lines.append("-- ============================================================")
        lines.append("")
        for ao_name in sorted(asset_owners):
            ao_slug = generate_slug(ao_name)
            lines.append(f"INSERT INTO asset_owners (name, slug)")
            lines.append(f"SELECT {_sql_str(ao_name)}, {_sql_str(ao_slug)}")
            lines.append(f"WHERE NOT EXISTS (SELECT 1 FROM asset_owners WHERE slug = {_sql_str(ao_slug)});")
            lines.append("")

    # Step 4: Insert developments with FK lookups
    lines.append("-- ============================================================")
    step = 3 if asset_owners else (2 if operators else 1)
    lines.append(f"-- Step {step}: Insert developments")
    lines.append("-- ============================================================")
    lines.append("")

    for dev in eligible:
        lines.append(f"-- {dev.name} ({dev.area or 'unknown area'})")
        lines.append(f"-- Confidence: {dev.confidence.value} ({dev.confidence_score:.2f})")
        lines.append(f"-- Sources: {', '.join(dev.source_urls[:3])}")

        # Build DO $$ block if we need FK lookups
        needs_fk = bool(dev.operator_name or dev.asset_owner_name)

        if needs_fk:
            lines.append("DO $$")
            lines.append("DECLARE")
            if dev.operator_name:
                lines.append("  op_id UUID;")
            if dev.asset_owner_name:
                lines.append("  ao_id UUID;")
            lines.append("BEGIN")

            if dev.operator_name:
                op_slug = generate_slug(dev.operator_name)
                lines.append(f"  SELECT id INTO op_id FROM operators WHERE slug = {_sql_str(op_slug)};")
            if dev.asset_owner_name:
                ao_slug = generate_slug(dev.asset_owner_name)
                lines.append(f"  SELECT id INTO ao_id FROM asset_owners WHERE slug = {_sql_str(ao_slug)};")

            lines.append("")
            lines.append(f"  INSERT INTO developments (")
            fields, values = _build_insert_fields(dev, use_fk_vars=True)
            lines.append(f"    {', '.join(fields)}")
            lines.append(f"  ) SELECT")
            lines.append(f"    {', '.join(values)}")
            lines.append(f"  WHERE NOT EXISTS (SELECT 1 FROM developments WHERE slug = {_sql_str(dev.slug)});")
            lines.append("END $$;")
        else:
            lines.append(f"INSERT INTO developments (")
            fields, values = _build_insert_fields(dev, use_fk_vars=False)
            lines.append(f"  {', '.join(fields)}")
            lines.append(f") SELECT")
            lines.append(f"  {', '.join(values)}")
            lines.append(f"WHERE NOT EXISTS (SELECT 1 FROM developments WHERE slug = {_sql_str(dev.slug)});")

        lines.append("")

    with open(filepath, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))

    return filepath


def _build_insert_fields(dev: DiscoveredDevelopment, use_fk_vars: bool) -> tuple[list[str], list[str]]:
    """Build the field names and values for a development INSERT."""
    fields = ["name", "slug", "development_type"]
    values = [_sql_str(dev.name), _sql_str(dev.slug), _sql_str(dev.development_type)]

    if dev.operator_name:
        fields.append("operator_id")
        values.append("op_id" if use_fk_vars else "NULL")

    if dev.asset_owner_name:
        fields.append("asset_owner_id")
        values.append("ao_id" if use_fk_vars else "NULL")

    if dev.area:
        fields.append("area")
        values.append(_sql_str(dev.area))

    if dev.region:
        fields.append("region")
        values.append(_sql_str(dev.region))

    if dev.postcode:
        fields.append("postcode")
        values.append(_sql_str(dev.postcode))

    if dev.latitude is not None:
        fields.append("latitude")
        values.append(str(dev.latitude))

    if dev.longitude is not None:
        fields.append("longitude")
        values.append(str(dev.longitude))

    if dev.number_of_units:
        fields.append("number_of_units")
        values.append(str(dev.number_of_units))

    if dev.status:
        fields.append("status")
        values.append(_sql_str(dev.status))

    if dev.completion_date:
        fields.append("completion_date")
        values.append(_sql_str(dev.completion_date))

    if dev.description:
        fields.append("description")
        values.append(_sql_str(dev.description[:500]))

    if dev.website_url:
        fields.append("website_url")
        values.append(_sql_str(dev.website_url))

    return fields, values


def _sql_str(val: str) -> str:
    """SQL-safe string literal."""
    escaped = val.replace("'", "''")
    return f"'{escaped}'"
